app/com/arpnetworking/rollups/RollupGenerator.java

```diff
@@ -185,41 +222,45 @@ private void handleTagNamesMessage(final TagNamesMessage message) {
 
             getSelf().tell(
                     new FinishRollupMessage.Builder()
-                            .setMetricName(message.getMetricName())
-                            .setPeriod(message.getPeriod())
-                            .setFailure(message.getFailure().orElse(new RuntimeException("Received Failure")))
+                            .setMetricName(metricName)
+                            .setPeriod(period)
+                            .setFailure(throwable)
                             .build(),
                     ActorRef.noSender());
         } else {
             _metrics.recordCounter("rollup/generator/last_data_point_message/success", 1);
-            final Instant recentPeriodStartTime = message.getPeriod().recentStartTime(_clock.instant());
+            final Instant startOfRecentClosedPeriod = period.recentStartTime(_clock.instant());
+            final Instant lastDataPoint = message.getLastDataPointTime().orElse(Instant.EPOCH);
 
             // If the most recent period aligned start time is after the most recent datapoint then
             // we need to run the rollup, otherwise we can skip this and just send a finish message.
-            if (recentPeriodStartTime.isAfter(message.getLastDataPointTime().orElse(Instant.EPOCH))) {
-                final String rollupMetricName = message.getMetricName() + message.getPeriod().getSuffix();
-                final RollupPeriod period = message.getPeriod();
+            if (startOfRecentClosedPeriod.isAfter(lastDataPoint)) {
+                final String rollupMetricName = metricName + period.getSuffix();
+                final int maxBackFillPeriods = _maxBackFillByPeriod.getOrDefault(period, 0);
 
-                // The last datapoint contains data for the period that follows it
-                final Instant lastDataPoint = message.getLastDataPointTime().orElse(Instant.EPOCH);
-                final Instant oldestBackfillPoint = period.recentEndTime(
-                        _clock.instant()).minus(period.periodCountToDuration(_maxBackFillPeriods));
-                final Instant startOfRecentClosedPeriod = period.recentStartTime(_clock.instant());
+                final Instant oldestBackfillPoint = period.recentEndTime(_clock.instant())
+                        .minus(period.periodCountToDuration(maxBackFillPeriods));
 
-                // We either want to start at the oldest backfill point or the start of the period after the last datapoint.
+                // We either want to start at the oldest backfill point or the start of the period
+                // after the last datapoint since it contains data for the period that follows it.
                 Instant rollupPeriodStart = lastDataPoint.isBefore(oldestBackfillPoint)
-                        ? oldestBackfillPoint : lastDataPoint.plus(period.periodCountToDuration(1));
+                        ? oldestBackfillPoint : period.recentEndTime(lastDataPoint).plus(period.periodCountToDuration(1));
```

Comment:
This line had a bug where it didn't align the last data point to the most recent period. I fixed that and update the test.